### PR TITLE
[MachineLICM] Hoist copies of constant physical register

### PR DESCRIPTION
**Pull request by wangpc-pp**
Previously, we just check if the source is a virtual register and
this prevents some potential hoists.

We can see some improvements in AArch64/RISCV tests.

### CODE REVIEW

**Review by lukel97 — APPROVED** [review]
Makes sense to me and RISC-V diffs look good. Can't comment on what isConstantPhysReg implies for the other targets though so maybe wait for a 2nd reviewer

File: llvm/test/CodeGen/RISCV/rvv/fold-scalar-load-crash.ll [hunk 1]
```diff
@@ -12,21 +12,19 @@ define i32 @test(i32 %size, ptr %add.ptr, i64 %const) {
 ; RV32-NEXT:    vsetivli zero, 8, e8, mf2, ta, ma
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/rvv/fold-scalar-load-crash.ll:12]: Have we lost the intention of this test?

Comment (wangpc-pp) [reply llvm/test/CodeGen/RISCV/rvv/fold-scalar-load-crash.ll:12]: At least it doesn't crash? :-)
I will do some modifications to keep the loop.

**Comment by wangpc-pp** [comment]
Background: I'm doing some auto-vectorization evaluations of x264 hotspots on RISCV. This change can improve codegen vector reductions as we can hoist the initial zero value.

File: llvm/lib/CodeGen/MachineLICM.cpp [hunk 2]
```diff
@@ -1269,8 +1269,9 @@ bool MachineLICMBase::IsProfitableToHoist(MachineInstr &MI,
     Register DefReg = MI.getOperand(0).getReg();
     if (DefReg.isVirtual() &&
         all_of(MI.uses(),
-               [](const MachineOperand &UseOp) {
-                 return !UseOp.isReg() || UseOp.getReg().isVirtual();
+               [&](const MachineOperand &UseOp) {
+                 return !UseOp.isReg() || UseOp.getReg().isVirtual() ||
+                        MRI->isConstantPhysReg(UseOp.getReg());
```
Comment (arsenm) [inline llvm/lib/CodeGen/MachineLICM.cpp:1274]: MRI is only capture, 
```suggestion
               [=this](const MachineOperand &UseOp) {
                 return !UseOp.isReg() || UseOp.getReg().isVirtual() ||
                        MRI->isConstantPhysReg(UseOp.getReg());
```

**Review by arsenm — APPROVED** [review]

**Review by arsenm — APPROVED** [review]

**Pull request merged**